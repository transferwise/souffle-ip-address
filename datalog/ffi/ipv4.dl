// Operations assume netaddr in normalized form
.type Cidr = [ netaddr: unsigned, netmask: unsigned ]

.type IPv4 = unsigned

// See functors.cpp for implementation.
//
// 'stateful' is only needed since these work with complex types,
// otherwise the functions themselves are not stateful in the functional sense.

.functor parseCidr(cidr: symbol): Cidr stateful
// Bool-y return value
.functor cidrOverlaps(a: Cidr, b: Cidr): number stateful
.functor maskBits(c: Cidr): unsigned stateful

// Prints cidr in x.y.z.w/m format
.functor printCidr(cidr: Cidr): symbol stateful
