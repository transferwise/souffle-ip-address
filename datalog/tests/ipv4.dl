#include "../ffi/ipv4.dl"

.decl raw_cidr(c: symbol)

raw_cidr("192.168.0.1/32").
raw_cidr("192.168.0.0/16").
raw_cidr("10.0.0.0/8").
raw_cidr("10.0.0.0/10").
raw_cidr("10.1.0.0/10").

.decl cidr(c: Cidr)

cidr(@parseCidr(c)) :-
    raw_cidr(c).

.decl print_cidr(c: symbol)
.output print_cidr ()

print_cidr(@printCidr(c)) :-
    cidr(c).

.decl dominates(a: Cidr, b: Cidr)

dominates(a, b) :-
    cidr(a),
    cidr(b),
    @cidrOverlaps(a, b) != 0,
    @maskBits(a) < @maskBits(b).

.decl print_dominates(a: symbol, b: symbol)
.output print_dominates ()

print_dominates(@printCidr(a), @printCidr(b)) :-
   dominates(a, b).
