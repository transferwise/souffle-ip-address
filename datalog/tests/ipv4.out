---------------
print_dominates
a	b
===============
192.168.0.0/16	192.168.0.1/32
10.0.0.0/8	10.0.0.0/10
10.0.0.0/8	10.1.0.0/10
===============
---------------
print_cidr
c
===============
192.168.0.1/32
192.168.0.0/16
10.0.0.0/8
10.0.0.0/10
10.1.0.0/10
===============
